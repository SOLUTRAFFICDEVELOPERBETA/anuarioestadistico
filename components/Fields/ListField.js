import React from 'react';
import PropTypes from 'prop-types';
import styled from '@emotion/styled';
import {
    TextField,
    List,
    ListItem,
    ListItemText,
    ListItemSecondaryAction,
    IconButton,
    Tooltip,
    Box,
    Popover,
    Grid,
    MenuItem,
    InputAdornment,
    makeStyles
} from '@material-ui/core';
import { AddBoxTwoTone, Close, DeleteForever, DragHandle } from '@material-ui/icons';
import { DragPreviewImage, useDrag } from 'react-dnd';
import clsx from 'clsx';
import { grey } from '@material-ui/core/colors';
import { GRID_SIZES } from '../../constants/documents';
// import { useForm } from 'react-hook-form'
// import EditorContext from '../../contexts/editor'

// Componente contenedor de la lista
const ListContainer = styled(Box)`
    width: 100%;
    padding: 1rem;
    display: grid;
    grid-template-columns: 0px auto;

    :focus-within {
        background-color: #eee;
        z-index: 5;
        -webkit-box-shadow: 0px 0px 5px 0px rgba(0, 0, 0, 0.75);
        -moz-box-shadow: 0px 0px 5px 0px rgba(0, 0, 0, 0.75);
        box-shadow: 0px 0px 5px 0px rgba(0, 0, 0, 0.75);
    }

    :hover {
        grid-template-columns: 30px auto;
    }

    .dragger {
        align-self: center;
        justify-self: center;
        width: 24px;
        height: 24px;
        visibility: hidden;
        opacity: 0;
        transition: visibility 0s, opacity 0.5s linear;
    }

    :hover .dragger {
        visibility: visible;
        opacity: 1;
    }

    .dragger:hover {
        cursor: move;
    }
`;

// Campo del elemento
const ElementField = styled.div`
    width: 100%;
    display: inline-block;

    :hover {
        background-color: rgb(245, 245, 245);
    }

    :focus {
        border: 1px solid #ccc;
        border-radius: 5px;
        background-color: rgb(245, 245, 245);
        padding: 4px;
    }
`;

// Contenedor del elemento de la lista
const ElementItem = styled(ListItem)`
    .MuiListItemSecondaryAction-root {
        display: none;
    }

    :hover .MuiListItemSecondaryAction-root {
        display: block;
    }
`;

// Estilos del componente
const useStyles = makeStyles((theme) => ({
    dragging: {
        backgroundColor: '#eee',
        // z-index: 5,
        boxShadow: '0px 0px 5px 0px rgba(0,0,0,0.75)'
    },
    item: {
        [theme.breakpoints.up('md')]: {
            maxWidth: '50rem'
        },
        [theme.breakpoints.between('sm', 'md')]: {
            maxWidth: '25rem'
        }
    }
}));

/**
 * Componente que muestra una lista de elementos
 * @param {{ id: string, value: any, size: number | string, onChange: () => void, onDelete: () => void }} props Propiedades del componente
 */
const ListField = ({ id, value, size, onChange, onDelete }) => {
    const classes = useStyles();
    const [anchorEl, setAnchorEl] = React.useState(null);
    // const { onChangeField, onDeleteField } = React.useContext(EditorContext)
    const [element, setElement] = React.useState('');
    const [{ isDragging }, drag, preview] = useDrag({
        item: { type: 'ITEM', id },
        collect: (monitor) => ({
            isDragging: !!monitor.isDragging()
        })
    });

    /**
     * Método para abrir el popover
     * @param {Event} event Evento de click de la celda
     */
    const handleClick = (event) => {
        event.preventDefault();
        setAnchorEl(event.currentTarget);
    };

    // Método para cerrar el popover
    const handleClose = () => {
        setAnchorEl(null);
    };

    /**
     * Método de ingreso a la lista
     * @param {{ field: string }} values Campo del formulario
     */
    const onCreateElement = (event) => {
        event.preventDefault();
        // console.log({ element })
        if (element.trim() === '') return;

        const fields = [...value, element];
        setElement('');
        onChange({ value: fields });
    };

    /**
     * Método de cambio de los elementos
     * @param {Event} event Evento de cambio
     */
    const handleChangeElement = ({ target }) => {
        const fields = [...value].map((v, i) =>
            i.toString() === target.id ? target.textContent : v
        );
        onChange({ value: fields });
    };

    /**
     * Método de eliminación de la lista
     * @param {number} index Posición del elemento
     */
    const handleDeleteElement = (index) => {
        const fields = [...value].filter((_, _index) => _index !== index);
        onChange({ value: fields });
    };

    const open = Boolean(anchorEl);

    return (
        <React.Fragment>
            <DragPreviewImage connect={preview} src="/static/img/DragPreview.png" />
            <ListContainer
                ref={drag}
                padding={3}
                onContextMenu={handleClick}
                className={clsx({
                    [classes.dragging]: isDragging
                })}>
                <span className="dragger">
                    <DragHandle fontSize="small" style={{ fill: grey[500] }} />
                </span>
                <div>
                    <List dense>
                        {value.map((v, index) => (
                            <ElementItem key={index}>
                                <ListItemText
                                    primary={
                                        <ElementField
                                            contentEditable
                                            id={index}
                                            className={classes.item}
                                            onBlur={handleChangeElement}
                                            dangerouslySetInnerHTML={{ __html: v }}
                                        />
                                    }
                                />
                                <ListItemSecondaryAction>
                                    <IconButton
                                        size="small"
                                        onClick={() => handleDeleteElement(index)}>
                                        <Tooltip title="Borrar elemento" placement="left">
                                            <Close color="error" fontSize="small" />
                                        </Tooltip>
                                    </IconButton>
                                </ListItemSecondaryAction>
                            </ElementItem>
                        ))}
                    </List>
                    <form onSubmit={onCreateElement}>
                        <TextField
                            fullWidth
                            margin="dense"
                            size="small"
                            // variant="outlined"
                            name="field"
                            value={element}
                            placeholder="Añadir elemento"
                            onChange={(event) => setElement(event.target.value)}
                            InputProps={{
                                endAdornment: (
                                    <InputAdornment position="end">
                                        <IconButton type="submit">
                                            <AddBoxTwoTone color="primary" />
                                        </IconButton>
                                    </InputAdornment>
                                )
                            }}
                        />
                    </form>
                </div>
            </ListContainer>
            {open && (
                <Popover
                    id={`popover-open-${id}`}
                    open={open}
                    anchorEl={anchorEl}
                    onClose={handleClose}
                    anchorOrigin={{
                        vertical: 'top',
                        horizontal: 'left'
                    }}
                    transformOrigin={{
                        vertical: 'bottom',
                        horizontal: 'left'
                    }}>
                    <Box padding={1}>
                        <Grid container spacing={1} alignContent="center" alignItems="center">
                            <Grid item>
                                <TextField
                                    fullWidth
                                    label="Tamaño del objeto"
                                    select
                                    size="small"
                                    variant="outlined"
                                    value={size}
                                    name="size"
                                    style={{ minWidth: '160px' }}
                                    onChange={({ target: { value } }) => onChange({ size: value })}
                                    // className={classes.option}
                                >
                                    {GRID_SIZES.map(($size) => (
                                        <MenuItem
                                            disabled={size === $size.value}
                                            value={$size.value}
                                            key={`select-size-option-${id}-${$size.value}`}>
                                            {$size.label}
                                        </MenuItem>
                                    ))}
                                </TextField>
                            </Grid>
                            <Grid item>
                                <Tooltip title="Borrar" interactive>
                                    <IconButton size="small" onClick={() => onDelete(id)}>
                                        <DeleteForever color="error" />
                                    </IconButton>
                                </Tooltip>
                            </Grid>
                        </Grid>
                    </Box>
                </Popover>
            )}
        </React.Fragment>
    );
};

ListField.propTypes = {
    id: PropTypes.string.isRequired,
    size: PropTypes.any,
    value: PropTypes.array,
    onChange: PropTypes.func.isRequired,
    onDelete: PropTypes.func.isRequired
};

export default ListField;
