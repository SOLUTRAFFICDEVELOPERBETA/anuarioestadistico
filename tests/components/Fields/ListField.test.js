import '@testing-library/jest-dom';
import { act } from '@testing-library/react';
import { mount } from 'enzyme';
import React from 'react';
import { RouterContext } from 'next/dist/next-server/lib/router-context'
import { DndProvider } from 'react-dnd'
import { HTML5Backend } from 'react-dnd-html5-backend'
import AlertContext from '../../../contexts/alert';
import ListField from '../../../components/Fields/ListField';
import { valueTypeArray } from '../../fixtures';


describe('Pruebas en <ListField/>', () => {
    
    const onDelete = jest.fn();
    const onChange = jest.fn((data) => console.log(data));
    const showMessage = jest.fn();
    
    test('debe mostrarse correctamente', () => {
        const wrapper = mount(
            <DndProvider backend={HTML5Backend}>
                <AlertContext.Provider value={{
                    showMessage
                }}>
                    <ListField
                        id="id-ImageField"
                        value={valueTypeArray}
                        size={12}
                        onChange={onChange}
                        onDelete={onDelete}
                    />
                </AlertContext.Provider>
            </DndProvider>
        )
        expect(wrapper).toMatchSnapshot()
    });


    test('debe llamar la función onChange() ', () => {
        const wrapper = mount(
            <DndProvider backend={HTML5Backend}>
                <AlertContext.Provider value={{
                    showMessage
                }}>
                    <ListField
                        id="id-ImageField"
                        value={valueTypeArray}
                        size={12}
                        onChange={onChange}
                        onDelete={onDelete}
                    />
                </AlertContext.Provider>
            </DndProvider>
        )
        wrapper.find('div').at(4).prop('onBlur')({ target: { id: 1, textContent: 'lista 2 de prueba' } });

        expect(onChange).toHaveBeenCalled();
        expect(onChange).toHaveBeenCalledTimes(1);
        expect(onChange).toHaveBeenCalledWith(expect.any(Object));
        expect(onChange).toHaveBeenCalledWith({
            value: ['lista 1']
        });
    })

    test('debe de llamar la función onDelete()', () => {
        const wrapper = mount(
            <DndProvider backend={HTML5Backend}>
                <AlertContext.Provider value={{
                    showMessage
                }}>
                    <ListField
                        id="id-ImageField"
                        value={valueTypeArray}
                        size={12}
                        onChange={onChange}
                        onDelete={onDelete}
                    />
                </AlertContext.Provider>
            </DndProvider>
        )
        act(() => {
            wrapper.find('div').at(0).prop('onContextMenu')({ preventDefault() { }, currentTarget(){}});
        })
        // expect(wrapper).toMatchSnapshot();
        // expect(wrapper.find(''))
        // console.log(wrapper.html());
    })

})