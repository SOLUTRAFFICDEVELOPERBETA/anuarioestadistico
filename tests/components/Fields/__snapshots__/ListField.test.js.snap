// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <ListField/> debe mostrarse correctamente 1`] = `
Array [
  "",
  <div
    className="MuiBox-root MuiBox-root-3  css-1f81dgw-ListContainer e1y5bfzn0"
    onContextMenu={[Function]}
  >
    <span
      className="dragger"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
        focusable="false"
        style={
          Object {
            "fill": "#9e9e9e",
          }
        }
        viewBox="0 0 24 24"
      >
        <path
          d="M20 9H4v2h16V9zM4 15h16v-2H4v2z"
        />
      </svg>
    </span>
    <div>
      <ul
        className="MuiList-root MuiList-dense MuiList-padding"
      >
        <li
          className="MuiListItem-container"
        >
          <div
            className="MuiListItem-root css-y3qveq-ElementItem e1y5bfzn2 MuiListItem-dense MuiListItem-gutters MuiListItem-secondaryAction"
            disabled={false}
          >
            <div
              className="MuiListItemText-root MuiListItemText-dense"
            >
              <span
                className="MuiTypography-root MuiListItemText-primary MuiTypography-body2 MuiTypography-displayBlock"
              >
                <div
                  className="makeStyles-item-2 css-s64paw-ElementField e1y5bfzn1"
                  contentEditable={true}
                  dangerouslySetInnerHTML={
                    Object {
                      "__html": "lista 1",
                    }
                  }
                  id={0}
                  onBlur={[Function]}
                />
              </span>
            </div>
          </div>
          <div
            className="MuiListItemSecondaryAction-root"
          >
            <button
              className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall"
              disabled={false}
              onBlur={[Function]}
              onClick={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={0}
              type="button"
            >
              <span
                className="MuiIconButton-label"
              >
                Array [
                  <svg
                    aria-describedby={null}
                    aria-hidden={true}
                    className="MuiSvgIcon-root MuiSvgIcon-colorError MuiSvgIcon-fontSizeSmall"
                    focusable="false"
                    onBlur={[Function]}
                    onFocus={[Function]}
                    onMouseLeave={[Function]}
                    onMouseOver={[Function]}
                    onTouchEnd={[Function]}
                    onTouchStart={[Function]}
                    title="Borrar elemento"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                    />
                  </svg>,
                  "",
                ]
              </span>
              <span
                className="MuiTouchRipple-root"
              />
            </button>
          </div>
        </li>
      </ul>
      <form
        onSubmit={[Function]}
      >
        <div
          className="MuiFormControl-root MuiTextField-root MuiFormControl-marginDense MuiFormControl-fullWidth"
        >
          <div
            className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedEnd MuiInputBase-marginDense MuiInput-marginDense"
            onClick={[Function]}
          >
            <input
              aria-invalid={false}
              autoFocus={false}
              className="MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedEnd MuiInputBase-inputMarginDense MuiInput-inputMarginDense"
              disabled={false}
              name="field"
              onAnimationStart={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              placeholder="AÃ±adir elemento"
              required={false}
              type="text"
              value=""
            />
            <div
              className="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-marginDense"
            >
              <button
                className="MuiButtonBase-root MuiIconButton-root"
                disabled={false}
                onBlur={[Function]}
                onDragLeave={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                tabIndex={0}
                type="submit"
              >
                <span
                  className="MuiIconButton-label"
                >
                  <svg
                    aria-hidden={true}
                    className="MuiSvgIcon-root MuiSvgIcon-colorPrimary"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M5 19h14V5H5v14zm2-8h4V7h2v4h4v2h-4v4h-2v-4H7v-2z"
                      opacity=".3"
                    />
                    <path
                      d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zm0 16H5V5h14v14zm-8-2h2v-4h4v-2h-4V7h-2v4H7v2h4z"
                    />
                  </svg>
                </span>
                <span
                  className="MuiTouchRipple-root"
                />
              </button>
            </div>
          </div>
        </div>
      </form>
    </div>
  </div>,
]
`;
